### PR TITLE
implement/fix reverse iteration when iterating with ending_before

### LINKED ISSUE #1090: allow autoPaginate with ending_before
hello, I'm experiencing the same issue in node as this has been reported here

https://github.com/stripe/stripe-dotnet/issues/1848

could you also fix this in the node version?

ty

### PR DESCRIPTION
**Pull request by kaznovac**
fixes #1090 

implementation, similarly to stripe/stripe-dotnet#1848, reverses the order of the response list to prevent unexpected iteration
(example without reversing described in https://github.com/stripe/stripe-dotnet/pull/1854/files#r350498976)

### CODE REVIEW

**Comment by mnooblet** [comment]
nice, can we get this merged please?

**Comment by remi-stripe** [comment]
We'll get to it in early January after the holidays!

**Comment by mnooblet** [comment]
> We'll get to it in early January after the holidays!

great, happy new year  ;)

File: lib/autoPagination.js [hunk 1]
```diff
@@ -21,6 +22,10 @@ function makeAutoPaginationMethods(self, requestArgs, spec, firstPagePromise) {
       );
     }
 
+    if (i === 0 && reverseIteration) {
+      listResult.data.reverse();
```
Comment (richardm-stripe) [inline lib/autoPagination.js:26]: We should avoid mutating the list. Can use "reverseIteration" to index from the end, instead? The mutation can leak in subtle ways, e.g.:
```javacript
      const req = stripe.customers.list({
        limit: 2,
        email,
        ending_before: midPointCustomerId,
      });
      const firstTwoCusts = await req;
      console.log(firstTwoCusts.data.map((x) => x.id)); // cus_3, cus_4
      const moreCusts = await req.autoPagingToArray({limit: 4});
      console.log(moreCusts.map((x) => x.id)); // cus_4, cus_3, cus_2, cus_1
      console.log(firstTwoCusts.data.map((x) => x.id)); // cus_4, cus_3
```

**Review by richardm-stripe — CHANGES_REQUESTED** [review]
Hi @kaznovac, very grateful for this PR! Thank you especially for the tests. I just have one request to avoid `.reverse()`.

**Comment by richardm-stripe** [comment]
Wrote a couple of extra tests while reviewing this to make sure I understood this change. You can merge them into your branch if you'd like, otherwise I will merge it afterwards:

https://github.com/kaznovac/stripe-node/compare/reverse-iteration...richardm-stripe:richardm-reverse-iteration

Lots of whitespace noise because I added a `describe` unfortunately, but the new tests are

https://github.com/kaznovac/stripe-node/compare/reverse-iteration...richardm-stripe:richardm-reverse-iteration#diff-4e53030d8c08ab9ded8e272ce6840dec73ec6d10d746382502f274378da2a474R529-R542

and

https://github.com/kaznovac/stripe-node/compare/reverse-iteration...richardm-stripe:richardm-reverse-iteration#diff-4e53030d8c08ab9ded8e272ce6840dec73ec6d10d746382502f274378da2a474R580-R697

File: lib/autoPagination.js (hunk 1 above)
Comment (kaznovac) [reply lib/autoPagination.js:26]: @richardm-stripe thanks for the review, and much thanks for pointing this out (it didn't occur to me)

i'll see to implement this - thinking of closure around the index to save us the conditional-branching on the `reverseIteration` on the each iteration

Comment (kaznovac) [reply lib/autoPagination.js]: made a simple numeric index manipulation, as suggested (91c78e9); adding another array iterator closure feels redundant and awkwardly similar to existing `iterate` function.

**Review by richardm-stripe — APPROVED** [review]

**Comment by richardm-stripe** [comment]
Thank you so much for the contribution @kaznovac, we'll get this released shortly.

**Comment by mnooblet** [comment]
@kaznovac @richardm-stripe thank you guys 🍺

**Pull request merged**